### ISSUE TITLE
Assigning a nn.Module to a property overwrites the property, rather than using its setter

### ISSUE THREAD

**Issue by SamNPowers** [pytorch/pytorch#52664]
## 🐛 Bug

I would like to be able to use properties in my Modules, as a convenient way to intercept assignment and do something slightly different. However, if I try to have a property that holds Modules, not only is the setter is completely bypassed, but the property is completely overwritten by the Module.

The code pathway that explains why this happens is discussed here: [https://stackoverflow.com/questions/61116433/maybe-i-found-something-strange-on-pytorch-which-result-in-property-setter-not](https://stackoverflow.com/questions/61116433/maybe-i-found-something-strange-on-pytorch-which-result-in-property-setter-not). However I disagree with the answerer that this is not a bug, as this is quite unexpected. In many use cases the failure will be silent: a code path that works well for non-Module objects will still seem to work for Modules (as the assigned Module is accessible via the property name), but whatever else custom you wanted simply won't occur.

## To Reproduce
```
import torch.nn as nn


class TestClass(nn.Module):
    def __init__(self):
        super().__init__()
        self._some_property = None
        self.indicator = "not set"

    @property
    def some_property(self):
        return self._some_property

    @some_property.setter
    def some_property(self, value):
        self.indicator = "set"
        self._some_property = value


if __name__ == "__main__":
    foo = TestClass()
    foo.some_property = "not a module"
    print(f"Result foo: {foo.indicator}")

    bar = TestClass()
    bar.some_property = nn.Module()
    print(f"Result bar: {bar.indicator}")

```
The above code is a pretty standard property definition plus a simple side effect that indicates whether the setter was actually called. For a non-Module object being assigned, it works as expected, with the indicator getting set. However for a Module, we can see the setter was bypassed.

[Observed]:
```
Result foo: set
Result bar: not set
```

[Expected]:
```
Result foo: set
Result bar: set
```

Note that if you change `class TestClass(nn.Module)` to `class TestClass(object)`, we get the expected result.

## Environment

PyTorch version: 1.4.0 (I glanced at latest, and the code path seems the same)
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 10.15.7 (x86_64)
GCC version: Could not collect
Clang version: 12.0.0 (clang-1200.0.32.29)
CMake version: version 3.18.0

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] numpydoc==1.1.0
[pip3] torch==1.4.0
[pip3] torch-ac==1.1.0
[pip3] torchaudio==0.4.0a0+719bcc7
[pip3] torchvision==0.5.0
[conda] Could not collect

## Additional thoughts

My two cents is that what I expect to happen is for the setter to be called, and it's the assignment within the setter that triggers adding the Module to the modules list. I think this wouldn't be too hard a change (perhaps just checking in  \_\_setattr\_\_  if `name` is a property on the object, and if so do the "normal" thing), though it does seem to be in a very significant, highly called code path.

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

**Comment by SamNPowers** [comment]
Putting this in the module I want to use parameters on seems to work so far:
```
def __setattr__(self, name, value):
        if name in vars(type(self)) and isinstance(vars(type(self))[name], property):
            object.__setattr__(self, name, value)
        else:
            super().__setattr__(name, value)
```

**Comment by ailzhang** [comment]
Putting this in triage review since I don't know which label to assign here. The bug report looks valid to me.

**Comment by albanD** [comment]
I am not sure we can do much here.
We need to intercept setting nn.Module to be able to properly register them.
Note that you will see the same behavior if you use nn.Parameter or regular Tensors after setting that attribute as a buffer.

Note that your custom `__setattr__` might prevent the Module from being properly registered and it won't be discovered properly when you move things to a different device or get the paramters!

**Comment by SamNPowers** [comment]
Well the idea I had been thinking is basically:
The first time \_\_setattr\_\_ is called, it's a property, and it goes through the custom handler, calling the property function (no modules assigned, as expected so far).
Within the property function, which looks something like:
```
@my_val.setter
def my_val(self, new val):
    self._my_val = new_val
```
when self.\_my\_val is set (which is not a property), it goes through the normal path, and \_my\_val gets properly registered. (If the setter looks totally different and no actual object variable is being set, I wouldn't even expect to go through the registration path anyhow.) I wouldn't really expect anything to be registered for just setting the property itself (since it's really only a function call as-is), only if something on the class is actually set inside that function would I expect it.

And if no setter is set, the user gets an AttributeError if they try to set the property anyway (one of the key benefits of using a property, in my opinion). So the no-custom-setter-case is also handled by this method. 

In other words, Modules do get properly registered in this flow, as far as I can tell.

**Comment by SamNPowers** [comment]
BTW my original \_\_setattr\_\_ didn't handle multiple inheritance quite right. Here's an updated version that looks at all of the base classes as well, for whether the property is on any of them.

```
    def __setattr__(self, name, value):
        # Explicitly get vars from all base classes (via the method resolution attribute __mro__)
        all_vars = {}
        for base_class in type(self).__mro__:
            all_vars.update(vars(base_class))

        if name in all_vars and isinstance(all_vars[name], property):
            object.__setattr__(self, name, value)
        else:
            super().__setattr__(name, value)
```

**Comment by saitcakmak** [comment]
Would it be possible to raise an error or warning if attempting to set a Module for an attribute that defines a setter? The current behavior silently ignores the property setter and can cause serious bugs since the assignment effectively gets ignored. 

```
from torch.nn import Module

class Foo(Module):
    def __init__(self, bar = None):
        super().__init__()
        self._bar = bar

    @property
    def bar(self):
        return self._bar
    
    @bar.setter
    def bar(self, bar):
        self._bar = bar


foo = Foo()
foo.bar = Module()
```
The result of this operation is that `foo.bar` is `None`. `foo._modules` is `OrderedDict([('bar', Module())])` but that module is not accessible.

**Comment by albanD** [comment]
This is related to https://github.com/pytorch/pytorch/issues/49726
We would definitely accept a PR raising an error for this. Or even better, make it work properly!

**Comment by function2-llx** [comment]
I'm sorry to bother this issue that was closed one year ago, I just want to leave a hint for anyone come here.

The proposed PR was reverted due to "Caused regressions in a Meta internal model". Currently, there is still no error raises for property setter of `nn.Module` nor it works properly.

**Comment by arthurmccray** [comment]
Just dropping in to say that I have also ran into this, and raising an error or at least not failing silently would be helpful
